### PR TITLE
feat(metric): add some metrics for controllers

### PR DESCRIPTION
**Pull request by wy-lucky**
1. NamespaceAutoProp Controller:
# HELP kubeadmiral_controller_manager_namespace_propagate_failed_total
# TYPE kubeadmiral_controller_manager_namespace_propagate_failed_total gauge
kubeadmiral_controller_manager_namespace_propagate_failed_total{namespace="wy-namespaces"} 1

2. PolicyRC Controller: 
# HELP kubeadmiral_controller_manager_propagation_policy_reference_count 
# TYPE kubeadmiral_controller_manager_propagation_policy_reference_count gauge
kubeadmiral_controller_manager_propagation_policy_reference_count{name="default/wy",namespace="default"} 2

3. Status Controller:
# HELP kubeadmiral_controller_manager_status_collection_duration_seconds_seconds
# TYPE kubeadmiral_controller_manager_status_collection_duration_seconds_seconds summary
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.5"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.95"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.99"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds_sum{cluster="kubeadmiral-member-1",name="wy-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment"} 0.000177742
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds_count{cluster="kubeadmiral-member-1",name="wy-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment"} 5
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy123-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.5"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy123-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.95"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds{cluster="kubeadmiral-member-1",name="wy123-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment",quantile="0.99"} NaN
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds_sum{cluster="kubeadmiral-member-1",name="wy123-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment"} 0.000124493
kubeadmiral_controller_manager_status_collection_duration_seconds_seconds_count{cluster="kubeadmiral-member-1",name="wy123-deployments.apps",namespace="default",resource_gvk="apps/v1, Kind=Deployment"} 4

### CODE REVIEW

File: pkg/controllers/nsautoprop/controller.go [hunk 1]
```diff
@@ -255,6 +255,8 @@ func (c *Controller) reconcile(ctx context.Context, qualifiedName common.Qualifi
 		return worker.StatusAllOK
 	}
 
+	c.recordNsAutoPropError(fedNamespace)
```
Comment (limhawjia) [inline pkg/controllers/nsautoprop/controller.go:258]: what exactly are we trying to measure here? this only records the sync failures of namespace objects. if that's the case, would a corresponding metric emitted by the sync controller suffice?

Comment (wy-lucky) [reply pkg/controllers/nsautoprop/controller.go:258]: yes, only records the sync failures of namespace objects.

Comment (wy-lucky) [reply pkg/controllers/nsautoprop/controller.go:258]: other resources may not have this requirement.

File: pkg/controllers/policyrc/controller.go [hunk 2]
```diff
@@ -317,5 +317,14 @@ func (c *Controller) reconcilePersist(
 		}
 	}
 
+	if qualifiedName.Namespace == "" {
+		c.metrics.Store("cluster_"+metricName, newRefCount, stats.Tag{Name: "name", Value: qualifiedName.String()})
```
Comment (mrlihanbo) [inline pkg/controllers/policyrc/controller.go:321]: why we need a cluster prefix here?

File: pkg/controllers/status/controller.go [hunk 3]
```diff
@@ -643,6 +652,16 @@ func (s *StatusController) clusterStatuses(
 	return clusterStatus
 }
 
+func (s *StatusController) recordClusterStatusError(name, namespace, resourceGvk, cluster, errorType string) {
```
Comment (mrlihanbo) [inline pkg/controllers/status/controller.go:655]:
```suggestion
func (s *StatusController) recordStatusCollectionError(name, namespace, resourceGvk, cluster, errorType string) {
```

Comment (wy-lucky) [reply pkg/controllers/status/controller.go]: done

File: pkg/controllers/policyrc/controller.go (hunk 2 above)
Comment (wy-lucky) [reply pkg/controllers/policyrc/controller.go]: done

File: pkg/controllers/nsautoprop/controller.go [hunk 4]
```diff
@@ -396,3 +398,17 @@ func (c *Controller) HasSynced() bool {
 		c.namespaceInformer.Informer().HasSynced() &&
 		c.informerManager.HasSynced()
 }
+
+func (c *Controller) recordNsAutoPropError(fedNamespace *fedcorev1a1.ClusterFederatedObject) {
```
Comment (mrlihanbo) [inline pkg/controllers/nsautoprop/controller.go:402]: RecordNamespacePropagateFailedMetric. /cc @limhawjia

File: pkg/controllers/policyrc/controller.go [hunk 5]
```diff
@@ -317,5 +317,10 @@ func (c *Controller) reconcilePersist(
 		}
 	}
 
+	c.metrics.Store(metricName, newRefCount, []stats.Tag{
+		{Name: "name", Value: qualifiedName.String()},
```
Comment (mrlihanbo) [inline pkg/controllers/policyrc/controller.go:321]:
```suggestion
		{Name: "name", Value: qualifiedName.Name},
```

File: pkg/controllers/nsautoprop/controller.go (hunk 4 above)
Comment (limhawjia) [reply pkg/controllers/nsautoprop/controller.go:402]: `emitNamespacePropagationFailedMetric`?

Comment (mrlihanbo) [reply pkg/controllers/nsautoprop/controller.go:402]: lgtm

File: pkg/controllers/status/controller.go [hunk 6]
```diff
@@ -643,6 +652,16 @@ func (s *StatusController) clusterStatuses(
 	return clusterStatus
 }
 
+func (s *StatusController) recordStatusCollectionError(name, namespace, resourceGvk, cluster, errorType string) {
+	s.metrics.Counter("status_collection_error_total", 1, []stats.Tag{
+		{Name: "name", Value: name},
+		{Name: "namespace", Value: namespace},
+		{Name: "resource_gvk", Value: resourceGvk},
```
Comment (mrlihanbo) [inline pkg/controllers/status/controller.go:659]:
```suggestion
		{Name: "resource", Value: resourceGvk},
```
/cc @limhawjia

File: pkg/controllers/status/controller.go [hunk 7]
```diff
@@ -571,6 +572,7 @@ func (s *StatusController) clusterStatuses(
 			resourceClusterStatus.Error = errMsg
 			clusterStatus = append(clusterStatus, resourceClusterStatus)
 			errList = append(errList, fmt.Sprintf("cluster-name: %s, error-info: %s", clusterName, errMsg))
+			s.recordStatusCollectionError(fedObject.GetName(), fedObject.GetNamespace(), targetGVK.String(), clusterName, errMsg)
```
Comment (mrlihanbo) [inline pkg/controllers/status/controller.go:575]: maybe err.Error() is more suitable here？ /cc @limhawjia

File: pkg/controllers/status/controller.go (hunk 6 above)
Comment (limhawjia) [reply pkg/controllers/status/controller.go:659]: lgtm

File: pkg/controllers/status/controller.go (hunk 7 above)
Comment (limhawjia) [reply pkg/controllers/status/controller.go:575]: would both `err.Error()` and `errMsg` be too verbose for a metric tag?

File: pkg/controllers/status/controller.go (hunk 6 above)
Comment (wy-lucky) [reply pkg/controllers/status/controller.go]: done

File: pkg/controllers/policyrc/controller.go (hunk 5 above)
Comment (wy-lucky) [reply pkg/controllers/policyrc/controller.go]: done

File: pkg/controllers/nsautoprop/controller.go (hunk 4 above)
Comment (wy-lucky) [reply pkg/controllers/nsautoprop/controller.go]: done

File: pkg/controllers/status/controller.go (hunk 7 above)
Comment (limhawjia) [reply pkg/controllers/status/controller.go:575]: let's remove this tag, the error is returned from the underlying informer store and would not be too meaningful

Comment (wy-lucky) [reply pkg/controllers/status/controller.go]: done

File: pkg/controllers/status/controller.go [hunk 8]
```diff
@@ -571,6 +572,7 @@ func (s *StatusController) clusterStatuses(
 			resourceClusterStatus.Error = errMsg
 			clusterStatus = append(clusterStatus, resourceClusterStatus)
 			errList = append(errList, fmt.Sprintf("cluster-name: %s, error-info: %s", clusterName, errMsg))
+			s.recordStatusCollectionError(fedObject.GetName(), fedObject.GetNamespace(), targetGVK.String(), clusterName)
```
Comment (mrlihanbo) [inline pkg/controllers/status/controller.go:575]: should use source object name

File: pkg/controllers/status/controller.go [hunk 9]
```diff
@@ -627,6 +630,12 @@ func (s *StatusController) clusterStatuses(
 			)
 		}
 		clusterStatus = append(clusterStatus, resourceClusterStatus)
+		s.metrics.Duration("status_collection_duration_seconds", startTime, []stats.Tag{
+			{Name: "name", Value: fedObject.GetName()},
```
Comment (mrlihanbo) [inline pkg/controllers/status/controller.go:634]: source object name

Comment (wy-lucky) [reply pkg/controllers/status/controller.go]: done

File: pkg/controllers/status/controller.go (hunk 8 above)
Comment (wy-lucky) [reply pkg/controllers/status/controller.go]: done

**Review by mrlihanbo — APPROVED** [review]

**Pull request merged**